### ISSUE TITLE
Noise in output

### ISSUE THREAD

**Issue by deepglugs** [lucidrains/imagen-pytorch#50]
Using the latest master, I'm noticing big improvement from 0.0.60.  The output form the upscaling unet isn't nearly as "swirly", but I am noticing red or green bits of noise on the output images:

(Top row unet1, second unet2)
![imagen_2_262](https://user-images.githubusercontent.com/70453896/173271916-25f1c1e7-a438-4973-bd6f-13d2909b859a.png)

I added a torch.clamp(0, 1.0) after the image is created in Imagen.sample(), but that didn't seem to help.  Any ideas where the noise is coming from?

**Comment by lucidrains** [comment]
cool! so i think it is also worth trying the old discrete version by setting https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/imagen_pytorch.py#L1231 to `False` and seeing if the artifacts are still there

i'm still slightly concerned about https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/imagen_pytorch.py#L248

**Comment by lucidrains** [comment]
also could try turning off the p2 loss https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/imagen_pytorch.py#L1232 for the continuous time case

**Comment by lucidrains** [comment]
@deepglugs thank you for reporting these results, and the artifacts!

**Comment by lucidrains** [comment]
@deepglugs actually may have identified the issue! let me know if the latest version fixes the problem https://github.com/lucidrains/imagen-pytorch/commit/ba50fc8c5dd8e651b87cab578cf7752bcc34ec91

**Comment by s4sarath** [comment]
Will there be too much difference between continuous time diffusion vs discrete times?

**Comment by deepglugs** [comment]
Kept continuous_times=True and updated to 0.1.14.  Changing p2_loss_weight_gamma=0 (and nothing else) doesn't seem to affect the noise.

Still has some noise, but it is more pronounced after the up-scaling but not always.

![imagen_28_18](https://user-images.githubusercontent.com/70453896/173383633-f11cc95f-df55-483c-ad91-ad72b96bef9e.png)

Could it have been something introduced between 0.0.60 and 0.1.10?  While these samples are not very good, there isn't any noise for the first unet:

![imagen_3_6666](https://user-images.githubusercontent.com/70453896/173384589-512075bd-5b9e-4e02-a621-55470139dded.png)

**Comment by lucidrains** [comment]
@deepglugs ohh as i feared, there is still some remaining issues with continuous time then :disappointed:  maybe i should default back to discrete ddpm

although it seems like you are getting good results with continuous time! it does appear to make a difference (although i'm not sure how much of it is due to p2 loss)

**Comment by lucidrains** [comment]
@deepglugs if you could run an experiment with `linear` only, that would help me narrow down if the issue is with continuous time or just the alpha cosine schedule in continuous time :pray:

**Comment by deepglugs** [comment]
How do I set `linear` only?  Will setting continous_times=False do that?

**Comment by lucidrains** [comment]
> How do I set `linear` only? Will setting continous_times=False do that?

you would just set `beta_schedules = ('linear',) * num_unets` when initializing `Imagen`

**Comment by ken012git** [comment]
I had this problem and I solve it by setting `use_linear_attn = False`.

```
unet1 = Unet(
        dim = 32,
        image_embed_dim = 512,
        num_resnet_blocks = 3,
        dim_mults=(1, 2, 3, 4),
        attn_dim_head = 64,
        layer_attns = (False, True, True, True), 
        use_linear_attn = False,
    )
```

I think the flag replaces transformers with linear attention layers at this [line](https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/imagen_pytorch.py#L993). However, I am not sure about which setting performs better.

**Comment by lucidrains** [comment]
@ken012git ohh no, those are two different things

`use_linear_attn` will supplement with linear attention, anywhere where quadratic full attention is not used

the `linear` i'm referring to is for the DDPM noise schedule

**Comment by deepglugs** [comment]
I think it might just be a clamping issue.  Before I mistakenly had ```torch.clamp(img, 0, 1.0)``` but it should have been ```img = torch.clamp(img, 0, 1.0)```

Trying this with use_linear_attn=False and True the noise is clear even with early samples:

![imagen_1_1000](https://user-images.githubusercontent.com/70453896/173446019-36d161cf-20a8-42cd-be0d-31ba53ce465f.png)

**Comment by lucidrains** [comment]
@deepglugs oh nice, yea that should be taken care of for you in the latest version https://github.com/lucidrains/imagen-pytorch/blob/main/imagen_pytorch/imagen_pytorch.py#L1445

**Comment by lucidrains** [comment]
@deepglugs and what i meant was `beta_schedules = ('linear', 'linear', 'linear')`

**Comment by lucidrains** [comment]
i should just rename `beta_schedules` to `noise_schedules` 🤔

**Comment by ken012git** [comment]
The loss curve seems strange when set `use_linear_attn = True` and will have noise in early results.

![unknown](https://user-images.githubusercontent.com/9960543/173448255-8aea3029-ca86-4a0c-a60c-91bc077b85d8.png)

<img width="471" alt="Screen_Shot_2022-06-12_at_11 13 46_AM" src="https://user-images.githubusercontent.com/9960543/173447966-89f4cf7d-23b8-4629-9462-52a690029923.png">

**Comment by lucidrains** [comment]
@ken012git ohh hmm, do you want to open a new issue for the linear attention issue?

**Comment by ken012git** [comment]
Sure, it's [here](https://github.com/lucidrains/imagen-pytorch/issues/55). Thanks!

**Comment by lucidrains** [comment]
@deepglugs when you turned off the p2 loss, do you see the same dramatic improvements you noted earlier? just wondering how much of it is due to continuous time vs the p2 loss reweighting

**Comment by lucidrains** [comment]
@deepglugs depending on how you answer that, i may also port the p2 loss weight over to the discrete time

**Comment by deepglugs** [comment]
> @deepglugs when you turned off the p2 loss, do you see the same dramatic improvements you noted earlier? just wondering how much of it is due to continuous time vs the p2 loss reweighting

I didn't run long enough to determine if it was much better.  I can do a longer run if you want once this test is complete.

**Comment by lucidrains** [comment]
@deepglugs yes, that'd be great, if you have the time! 🙏

**Comment by lucidrains** [comment]
@deepglugs i'm just wondering if continuous time is better than discrete or just an unimportant detail

**Comment by deepglugs** [comment]
Ran 6 epochs each.  First is default p2_loss, second is p2_loss=0:
![imagen_6_166](https://user-images.githubusercontent.com/70453896/173492764-73a48a8c-1d8a-4807-8703-12d925ca7788.png)
p2_loss=0:
![imagen_6_166](https://user-images.githubusercontent.com/70453896/173492784-9f4c2b0d-fe85-40bf-96e7-50342c45b34a.png)

**Comment by lucidrains** [comment]
@deepglugs thank you! i'll probably add the p2 loss as an option for discrete time gaussian diffusion tomorrow then!

**Comment by marunine** [comment]
For what it's worth, I've been getting good results on my dataset with `continuous_times=True` and `p2_loss_weight_gamma=1.0`.

I tend to get color shifting at the default of `0.5` until about my second epoch, whereas it starts converging within the first epoch at `1.0`. I think you can play around with it a bit as a hyperparameter per the paper where it comes from, but most of the experiments they ran seem to had it at `1.0`.

**Comment by lucidrains** [comment]
@marunine @deepglugs done! https://github.com/lucidrains/imagen-pytorch/commit/110855e11109c12c4c458d42d65399e872e7d0d3 thank you both for sharing your experimental results!
